security/pkg/nodeagent/caclient/providers/citadel/client_test.go

```diff
@@ -107,3 +115,108 @@ func TestCitadelClient(t *testing.T) {
+		return nil, fmt.Errorf("token is not valid")
+	}
+	return &pb.IstioCertificateResponse{CertChain: ca.Certs}, nil
+}
+
+func extractBearerToken(ctx context.Context) (string, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return "", fmt.Errorf("no metadata is attached")
+	}
+
+	authHeader, exists := md[authorizationMeta]
+	if !exists {
+		return "", fmt.Errorf("no HTTP authorization header exists")
+	}
+
+	for _, value := range authHeader {
+		if strings.HasPrefix(value, bearerTokenPrefix) {
+			return strings.TrimPrefix(value, bearerTokenPrefix), nil
+		}
+	}
+
+	return "", fmt.Errorf("no bearer token exists in HTTP authorization header")
+}
+
+// this test is to test whether the server side receive the correct token when
+// we build the CSR sign request
+func TestCitadelClientWithDifferentTypeToken(t *testing.T) {
+	testCases := map[string]struct {
+		server       mockTokenCAServer
+		expectedCert []string
+		expectedErr  string
+		token        string
+	}{
+		"Valid Token": {
+			server:       mockTokenCAServer{Certs: fakeCert},
+			expectedCert: fakeCert,
+			expectedErr:  "",
+			token:        validToken,
+		},
+		"Empty Token": {
+			server:       mockTokenCAServer{Certs: nil},
+			expectedCert: nil,
+			expectedErr:  "rpc error: code = Unknown desc = no HTTP authorization header exists",
+			token:        "",
+		},
+		"InValid Token": {
+			server:       mockTokenCAServer{Certs: []string{}},
+			expectedCert: nil,
+			expectedErr:  "rpc error: code = Unknown desc = token is not valid",
+			token:        fakeToken,
+		},
+	}
+
+	for id, tc := range testCases {
+		t.Run(id, func(t *testing.T) {
+			s := grpc.NewServer()
+			defer s.Stop()
+			lis, err := net.Listen("tcp", mockServerAddress)
+			if err != nil {
+				t.Fatalf("test case [%s]: failed to listen: %v", id, err)
+			}
+			go func() {
+				pb.RegisterIstioCertificateServiceServer(s, &tc.server)
+				if err := s.Serve(lis); err != nil {
+					t.Logf("Test case [%s]: failed to serve: %v", id, err)
+				}
+			}()
+
+			err = retry.UntilSuccess(func() error {
+				cli, err := NewCitadelClient(lis.Addr().String(), false, nil, "Kubernetes")
+				if err != nil {
+					return fmt.Errorf("test case [%s]: failed to create ca client: %v", id, err)
+				}
+				resp, err := cli.CSRSign(context.Background(), "12345678-1234-1234-1234-123456789012", []byte{01}, tc.token, 1)
+				if err != nil {
+					if err.Error() != tc.expectedErr {
+						return fmt.Errorf("test case [%s]: error (%s) does not match expected error (%s)", id, err.Error(), tc.expectedErr)
+					}
+				} else {
+					if tc.expectedErr != "" {
+						return fmt.Errorf("test case [%s]: expect error: %s but got no error", id, tc.expectedErr)
+					} else if !reflect.DeepEqual(resp, tc.expectedCert) {
+						return fmt.Errorf("test case [%s]: resp: got %+v, expected %v", id, resp, tc.expectedCert)
+					}
+				}
+				return nil
+			}, retry.Timeout(20*time.Second), retry.Delay(2*time.Second))
```

Comment:
Is the unit test super flaky? why do we need retry, usually we don't retry unless it's really flaky and there is no easy way to fix the flaky.

Reply:
Just prevent it to be flaky. As before this we sleep for one second for the server to ready. Here we wait for 2 seconds to waiting for server to be ready